src/tensor.cpp

```diff
@@ -0,0 +1,92 @@
+#include <tensor.hpp>
+#include <numeric>
+#include <algorithm>
+
+namespace deepworks {
+
+Tensor::Descriptor::Descriptor(const Shape &shape) : m_shape(shape) {
+    allocate(shape);
+}
+
+void Tensor::Descriptor::copyTo(Tensor::Descriptor &descriptor) {
+    if (this == &descriptor) {
+        throw std::runtime_error("Tensor cannot copy itself.");
+    }
+    if (m_shape != descriptor.m_shape || m_strides != descriptor.m_strides) {
+        throw std::runtime_error("Copy to another layout isn't supported.");
+    }
+
+    if (descriptor.m_data == nullptr) {
+        throw std::runtime_error("copyTo: Output tensor should be allocated.");
+    }
+
+    m_shape = descriptor.m_shape;
+    m_strides = descriptor.m_strides;
+
+    std::copy_n(m_data, m_total, descriptor.m_data);
+}
+
+void Tensor::Descriptor::allocate(const Shape &shape) {
+    if (m_data != nullptr) {
+        throw std::runtime_error("Tensor already allocated, cannot allocate twice.");
+    }
+    m_shape = shape;
+    m_total = std::accumulate(shape.begin(),
+                              shape.end(),
+                              1,
+                              std::multiplies<>());
+    m_data = new Type[m_total];
```

Comment:
What if shape is a vector = {-1}? Then m_total will be -1 also. You can`t allocate memory for negative size.